Error: GET https://api.github.com/repos/org/project/actions/secrets/xxxx: 404 Not Found []

I am trying to manage the Github secrets from terraform. I am facing various issues where during plan or delete I get 404 errors.

I tried during the delete to add --refresh=false or --refresh=true  but still getting the 

```hcl

## for a destroy:

 # module.github_secrets_aztfmod.github_actions_secret.ARM_SUBSCRIPTION_ID["aztfmod_level0"] will be destroyed
  - resource "github_actions_secret" "ARM_SUBSCRIPTION_ID" {
      - created_at      = "default" -> null
      - id              = "level0:ARM_SUBSCRIPTION_ID" -> null
      - plaintext_value = (sensitive value)
      - repository      = "level0" -> null
      - secret_name     = "ARM_SUBSCRIPTION_ID" -> null
      - updated_at      = "default" -> null
    }

---

module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_ID["aztfmod_level0"]: Destroying... [id=level0:ARM_CLIENT_ID]
module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_SECRET["aztfmod_level0"]: Destroying... [id=level0:ARM_CLIENT_SECRET]

Error: DELETE https://api.github.com/repos/aztfmod/level0/actions/secrets/ARM_TENANT_ID: 404  []

Error: DELETE https://api.github.com/repos/azure/caf-terraform-landingzones/actions/secrets/ARM_CLIENT_ID: 404  []

## for a plan
Error: GET https://api.github.com/repos/aztfmod/level0/actions/secrets/ARM_TENANT_ID: 404 Not Found []

Error: GET https://api.github.com/repos/azure/caf-terraform-landingzones/actions/secrets/ARM_CLIENT_SECRET: 404 Not Found []

```


```log
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: -----------------------------------------------------
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [TRACE] Acquiring lock for GitHub API request (%!q(<nil>))
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [DEBUG] Github API Request Details:
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: ---[ REQUEST ]---------------------------------------
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: GET /repos/azure/caf-terraform-landingzones/actions/secrets/ARM_TENANT_ID HTTP/1.1
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Host: api.github.com
2020-06-02T02:20:04.074Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: User-Agent: go-github
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Accept: application/vnd.github.v3+json
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Accept-Encoding: gzip
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: -----------------------------------------------------
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [TRACE] Acquiring lock for GitHub API request (%!q(<nil>))
2020-06-02T02:20:04.075Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [TRACE] Acquiring lock for GitHub API request (%!q(<nil>))
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [TRACE] Releasing lock for GitHub API request (%!q(<nil>))
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [DEBUG] Github API Response Details:
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: ---[ RESPONSE ]--------------------------------------
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: HTTP/1.1 404 Not Found
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Transfer-Encoding: chunked
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Access-Control-Allow-Origin: *
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Access-Control-Expose-Headers: ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, Deprecation, Sunset
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Content-Security-Policy: default-src 'none'
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Content-Type: application/json; charset=utf-8
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Date: Tue, 02 Jun 2020 02:20:04 GMT
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Referrer-Policy: origin-when-cross-origin, strict-origin-when-cross-origin
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Server: GitHub.com
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Status: 404 Not Found
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Strict-Transport-Security: max-age=31536000; includeSubdomains; preload
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: Vary: Accept-Encoding, Accept, X-Requested-With
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Accepted-Oauth-Scopes: 
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Content-Type-Options: nosniff
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Frame-Options: deny
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Github-Media-Type: github.v3; format=json
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Github-Request-Id: D2BD:0E6C:1E5484:282DA9:5ED5B753
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Oauth-Scopes: repo
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Ratelimit-Limit: 5000
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Ratelimit-Remaining: 4947
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Ratelimit-Reset: 1591066510
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: X-Xss-Protection: 1; mode=block
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 6b
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: {
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4:  "message": "Not Found",
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4:  "documentation_url": "https://developer.github.com/v3/actions/secrets/#get-a-secret"
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: }
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 0
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 
2020-06-02T02:20:04.321Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: -----------------------------------------------------
2020/06/02 02:20:04 [ERROR] module.github_secrets_aztfmod: eval: *terraform.EvalRefresh, err: GET https://api.github.com/repos/aztfmod/level0/actions/secrets/ARM_CLIENT_SECRET: 404 Not Found []
2020/06/02 02:20:04 [ERROR] module.github_secrets_aztfmod: eval: *terraform.EvalSequence, err: GET https://api.github.com/repos/aztfmod/level0/actions/secrets/ARM_CLIENT_SECRET: 404 Not Found []
2020/06/02 02:20:04 [TRACE] [walkRefresh] Exiting eval tree: module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_SECRET["aztfmod_level0"]
2020/06/02 02:20:04 [TRACE] vertex "module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_SECRET[\"aztfmod_level0\"]": visit complete
2020/06/02 02:20:04 [TRACE] vertex "module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_SECRET": dynamic subgraph encountered errors
2020/06/02 02:20:04 [TRACE] vertex "module.github_secrets_aztfmod.github_actions_secret.ARM_CLIENT_SECRET": visit complete
2020-06-02T02:20:04.422Z [DEBUG] plugin.terraform-provider-github_v2.8.0_x4: 2020/06/02 02:20:04 [TRACE] Releasing lock for GitHub API request (%!q(<nil>))
```

An easy way to reproduce the problem is to 
* create a secret using the resource github_actions_secret
* apply successfully
* manually destroy the secret from Github
* try apply and destroy

Expected behavior

If the secret does not exist it will be removed from the state. if refresh=true (default) 404 should not throw an error but use that information to cleanup the object from the state

during a plan /apply a 404 should be used to identity the resource has to be created.

--- a/github/resource_github_actions_secret.go
+++ b/github/resource_github_actions_secret.go
@@ -8,6 +8,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
 	"golang.org/x/crypto/nacl/box"
 	"log"
+	"net/http"
 )
 
 func resourceGithubActionsSecret() *schema.Resource {
@@ -101,7 +102,14 @@ func resourceGithubActionsSecretRead(d *schema.ResourceData, meta interface{}) e
 
 	secret, _, err := client.Actions.GetSecret(ctx, owner, repoName, secretName)
 	if err != nil {
-		d.SetId("")
+		if ghErr, ok := err.(*github.ErrorResponse); ok {
+			if ghErr.Response.StatusCode == http.StatusNotFound {
+				log.Printf("[WARN] Removing actions secret %s from state because it no longer exists in GitHub",
+					d.Id())
+				d.SetId("")
+				return nil
+			}
+		}
 		return err
 	}
 

--- a/github/resource_github_actions_secret_test.go
+++ b/github/resource_github_actions_secret_test.go
@@ -45,6 +45,28 @@ func TestAccGithubActionsSecret_basic(t *testing.T) {
 	})
 }
 
+func TestAccGithubActionsSecret_disappears(t *testing.T) {
+	repo := os.Getenv("GITHUB_TEMPLATE_REPOSITORY")
+	secretResourceName := "github_actions_secret.test_secret"
+	secretValue := "super_secret_value"
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckGithubActionsSecretDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccGithubActionsSecretFullConfig(repo, secretValue),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGithubActionsSecretExists(secretResourceName, "test_secret_name", t),
+					testAccCheckGithubActionsSecretDisappears(secretResourceName),
+				),
+				ExpectNonEmptyPlan: true,
+			},
+		},
+	})
+}
+
 func testAccGithubActionsSecretFullConfig(repoName, plaintext string) string {
 
 	// Take resources from other tests to avoid manual creation of secrets / repos
@@ -85,6 +107,23 @@ func testAccCheckGithubActionsSecretExists(resourceName, secretName string, t *t
 		}
 
 		return nil
+	}
+}
+
+func testAccCheckGithubActionsSecretDisappears(resourceName string) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[resourceName]
+		if !ok {
+			return fmt.Errorf("Not found: %s", resourceName)
+		}
+		conn := testAccProvider.Meta().(*Organization).v3client
+		owner := testAccProvider.Meta().(*Organization).name
+		repoName, secretName, err := parseTwoPartID(rs.Primary.ID, "repository", "secret_name")
+		if err != nil {
+			return err
+		}
+		_, err = conn.Actions.DeleteSecret(context.TODO(), owner, repoName, secretName)
+		return err
 	}
 }
 

--- a/github/resource_github_actions_secret_test.go
+++ b/github/resource_github_actions_secret_test.go
@@ -3,7 +3,7 @@ package github
 import (
 	"context"
 	"fmt"
-	"os"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
 	"testing"
 
 	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
@@ -11,7 +11,7 @@ import (
 )
 
 func TestAccGithubActionsSecret_basic(t *testing.T) {
-	repo := os.Getenv("GITHUB_TEMPLATE_REPOSITORY")
+	repo := acctest.RandomWithPrefix("tf-acc-test")
 
 	secretResourceName := "github_actions_secret.test_secret"
 	secretValue := "super_secret_value"
@@ -46,11 +46,11 @@ func TestAccGithubActionsSecret_basic(t *testing.T) {
 }
 
 func TestAccGithubActionsSecret_disappears(t *testing.T) {
-	repo := os.Getenv("GITHUB_TEMPLATE_REPOSITORY")
+	repo := acctest.RandomWithPrefix("tf-acc-test")
 	secretResourceName := "github_actions_secret.test_secret"
 	secretValue := "super_secret_value"
 
-	resource.Test(t, resource.TestCase{
+	resource.ParallelTest(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
 		Providers:    testAccProviders,
 		CheckDestroy: testAccCheckGithubActionsSecretDestroy,
@@ -68,22 +68,24 @@ func TestAccGithubActionsSecret_disappears(t *testing.T) {
 }
 
 func testAccGithubActionsSecretFullConfig(repoName, plaintext string) string {
-
-	// Take resources from other tests to avoid manual creation of secrets / repos
-	githubPKData := testAccCheckGithubActionsPublicKeyDataSourceConfig(repoName)
-	githubActionsSecretResource := testAccGithubActionsSecretConfig(repoName, plaintext)
-
-	return fmt.Sprintf("%s%s", githubPKData, githubActionsSecretResource)
+	// To allow tests to run in parallel and prevent re-using resources defined across the
+	// codebase, we create a repository resource and define it's actions public key here
+	// alongside the new actions secret resource
+	return fmt.Sprintf(`
+data "github_actions_public_key" "test_pk" {
+  repository = github_repository.test.name
 }
 
-func testAccGithubActionsSecretConfig(repo, plaintext string) string {
-	return fmt.Sprintf(`
+resource "github_repository" "test" {
+  name = "%s"
+}
+
 resource "github_actions_secret" "test_secret" {
-  repository       = "%s"
+  repository       = github_repository.test.name
   secret_name      = "test_secret_name"
   plaintext_value  = "%s"
 }
-`, repo, plaintext)
+`, repoName, plaintext)
 }
 
 func testAccCheckGithubActionsSecretExists(resourceName, secretName string, t *testing.T) resource.TestCheckFunc {

